relay-metrics/src/protocol.rs

```diff
@@ -152,20 +152,41 @@ fn is_valid_name(name: &str) -> bool {
     false
 }
 
-/// Validates an MRI of the form `<ty>:<ns>/<name>@<unit>`
-///
-/// Note that the format used in the statsd protocol is different: Metric names are not prefixed
-/// with `<ty>:` as the type is somewhere else in the protocol.
-pub(crate) fn is_valid_mri(name: &str) -> bool {
-    let mut components = name.splitn(2, ':');
-    if components.next().is_none() {
-        return false;
-    }
+/// A metric name parsed as MRI, a naming scheme which includes most of the metric's bucket key
+/// (excl. timestamp and tags).
+pub struct MetricMri {
+    /// The metric type.
+    pub ty: MetricType,
+    /// The namespace/usecase for this metric. For example `sessions` or `transactions`.
+    pub namespace: String,
+    /// The actual name, such as `duration` as part of `d:transactions/duration@ms`
+    pub name: String,
+    /// The metric unit.
+    pub unit: MetricUnit,
+}
```

Comment:
Rather than a structure for this, I'd prefer a newtype around a plain string that lazily parses fields of the MRI. During construction, it can still perform some early validation, or cache some results, though. 

Reasons for this is:
- It's easier to retain forward-compatibility, since unknown parts of the MRI will be forwarded, while still being able to parse the known parts. 
- This also applies to namespaces. You can create an enum and just return unknown, while automatically retaining the original value of the namespace.
- Albeit not important, fewer allocations

Reply:
>It's easier to retain forward-compatibility, since unknown parts of the MRI will be forwarded

A previous story implemented character validation in the metrics aggregator, which is enforced in every kind of Relay. I believe it's quite tricky to reconcile that design choice with what you're proposing. We need to decide whether we want to pass through garbage or not... everything we've been doing in terms of character validation over the past weeks has been heavily leaning towards the latter.